### ISSUE TITLE
Azure Account Sign-In does not work in VSCode when behind proxy

### ISSUE THREAD

**Issue by sathish331977** [microsoft/vscode-azure-account#27]
VSCode Version:1.8.1
OS Version: Windows 10 Professional
Steps to Reproduce:

View->Command Pallete-> Azure : Sign In
The popup to provide the device code to run in browser does not open. This results in not able to sign in to Azure account.
Try this without proxy. All works fine. Able to connect Azure account.

**Comment by chrmarti** [comment]
Does setting `HTTP_PROXY` and `HTTPS_PROXY` environment variables make it work?

**Comment by adriantorrie** [comment]
I have the same issue.

I have the environment variables set for my user account (as the corporate proxy requires username and password to get through). The env vars work when using other tools (conda, pip, etc).

Help > About output:
```
Version 1.19.3
Commit 7c4205b5c6e52a53b81c69d2b2dc8a627abaa0ba
Date 2018-01-25T10:36:34.867Z
Shell 1.7.9
Renderer 58.0.3029.110
Node 7.9.0
Architecture x64
```

**Comment by chrmarti** [comment]
@adriantorrie Which OS is this? Could you try starting VS Code from the command line making sure the env variables are set? (I don't have a proxy available, but setting HTTP_PROXY and HTTPS_PROXY to bogus values breaks the login.)

**Comment by adriantorrie** [comment]
OS: Windows 7

The env vars are set via my user profile, and are available with all terminal style applications. `HTTP_PROXY` & `HTTPS_PROXY` are set, I tried changing the names to `http_proxy` & `https_proxy` in case some weird case sensitivity was causing the issue. All other applications that rely on the env vars work as expected.

I've started from command line, still doesn't change anything (and just for completeness checked the env vars were set prior to launching). I also tried launching by explicitly passing in the via ``--proxy`` option, again, still doesn't fix anything.

I've also installed VS Code Insiders and run all the above steps. Still have the same issue.

I've also tried setting the proxy explicitly in the user settings via `"http.proxy"`. No change.

The thing is, for me to be able to install extensions from within the Extensions tab, VS Code needs to access the proxy, and this works fine and transparently.

It would seem it is an issue solely with the extension.

**Help > Toggle Developer Tools**

```sh
console.ts:123 [Extension Host] Error: Acquiring user code failed
	at C:\Users\torriea\.vscode-insiders\extensions\ms-vscode.azure-account-0.3.0\out\src\azure-account.js:555:28
	at Request._callback (C:\Users\torriea\.vscode-insiders\extensions\ms-vscode.azure-account-0.3.0\node_modules\adal-node\lib\util.js:119:7)
	at self.callback (C:\Users\torriea\.vscode-insiders\extensions\ms-vscode.azure-account-0.3.0\node_modules\request\request.js:188:22)
	at emitOne (events.js:96:13)
	at Request.emit (events.js:191:7)
	at Request.onRequestError (C:\Users\torriea\.vscode-insiders\extensions\ms-vscode.azure-account-0.3.0\node_modules\request\request.js:884:8)
	at emitOne (events.js:96:13)
	at ClientRequest.emit (events.js:191:7)
	at ClientRequest.onError (C:\Users\torriea\.vscode-insiders\extensions\ms-vscode.azure-account-0.3.0\node_modules\tunnel-agent\index.js:179:21)
	at Object.onceWrapper (events.js:293:19)
	at emitOne (events.js:96:13)
	at ClientRequest.emit (events.js:191:7)
	at TLSSocket.socketErrorListener (_http_client.js:358:9)
	at emitOne (events.js:96:13)
	at TLSSocket.emit (events.js:191:7)
	at onwriteError (_stream_writable.js:341:10)
	at onwrite (_stream_writable.js:359:5)
	at fireErrorCallbacks (net.js:480:13)
	at TLSSocket.Socket._destroy (net.js:521:3)
	at WriteWrap.afterWrite [as oncomplete] (net.js:815:10) (at originalConsole.(anonymous function) (c:\Program Files\Microsoft VS Code Insiders\resources\app\extensions\git\node_modules\diagnostic-channel-publishers\dist\src\console.pub.js:42:39))
t.log @ console.ts:123
```

**Comment by chrmarti** [comment]
@adriantorrie I get it to work for proxy basic auth by setting username and password in the proxy url:
```
export HTTPS_PROXY=http://username:password@proxy:8080
export HTTP_PROXY=http://username:password@proxy:8080
```

Could you retry and check what is logged to the console again? I improved logging in the last release.

**Comment by adriantorrie** [comment]
The proxy env vars I have set already have username and password in them, and are url encoded appropriately as well.

**Comment by chrmarti** [comment]
@adriantorrie Could you retry and check what is logged to the console again? I improved logging in the last release.

**Comment by satyakirank** [comment]
Im getting this error when trying to sign in azure account from vs code
 Error: getaddrinfo EAI_AGAIN login.microsoftonline.com:443
	at Object.exports._errnoException (util.js:1050:11)
	at errnoException (dns.js:33:15)
	at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:73:26)
console.ts:136 [Extension Host] Error: Acquiring user code failed
	at C:\Users\durgavenkata.neeli\.vscode\extensions\ms-vscode.azure-account-0.4.3\out\src\azure-account.js:550:28
	at Request._callback (C:\Users\durgavenkata.neeli\.vscode\extensions\ms-vscode.azure-account-0.4.3\node_modules\adal-node\lib\util.js:119:7)
	at self.callback (C:\Users\durgavenkata.neeli\.vscode\extensions\ms-vscode.azure-account-0.4.3\node_modules\request\request.js:186:22)
	at emitOne (events.js:96:13)
	at Request.emit (events.js:191:7)
	at Request.onRequestError (C:\Users\durgavenkata.neeli\.vscode\extensions\ms-vscode.azure-account-0.4.3\node_modules\request\request.js:878:8)
	at emitOne (events.js:96:13)
	at ClientRequest.emit (events.js:191:7)
	at TLSSocket.socketErrorListener (_http_client.js:358:9)
	at emitOne (events.js:96:13)
	at TLSSocket.emit (events.js:191:7)
	at connectErrorNT (net.js:1032:8)
	at _combinedTickCallback (internal/process/next_tick.js:80:11)
	at process._tickCallback (internal/process/next_tick.js:104:9)**

**Comment by chrmarti** [comment]
@satyakirank That looks like a separate problem. Could you open a new issue to track that?

**Comment by bhishma** [comment]
I have the similar issue as well. 
console.ts:136 [Extension Host] Error: connect ETIMEDOUT 65.52.1.21:443
	at Object._errnoException (util.js:1024:11)
	at _exceptionWithHostPort (util.js:1046:20)
	at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1182:14)

Error: Acquiring user code failed
	at C:\Users\sssdfsddf\.vscode\extensions\ms-vscode.azure-account-0.4.3\out\src\azure-account.js:550:28

**Comment by chrmarti** [comment]
@bhishma This looks like a separate problem, could you open a new issue to track it?

**Comment by revolz** [comment]
I am facing the same problem and same (temporary) solution works by having no proxy
The same description of the original post is still relevant, I just updated the Version to 1.29.0

--------------------------

VSCode Version:1.29.0
OS Version: Windows 10 Professional
Steps to Reproduce:

View->Command Pallete-> Azure : Sign In
The popup to provide the device code to run in browser does not open. This results in not able to sign in to Azure account.
Try this without proxy. All works fine. Able to connect Azure account.

**Comment by revolz** [comment]
Additional information, which might be relevant.

Visual Studio Code Tools for AI is unable to download sample also.
`verbose: Downloading sample 'MNIST using TensorFlow' ... failed, write EPROTO 101057795:error:14077419:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert access denied:d:\buildagent\workspace\318698\vendor\node\deps\openssl\openssl\ssl\s23_clnt.c:802:`


While other portion of Visual Studio Code (for example, extension download is working with the same proxy setting, C#, Python, even including Visual Studio Tools for AI extension can be downloaded without error).

I suspect it has the same root cause to the issue of this post. Hoping it will help the troubleshooting.

**Comment by revolz** [comment]
@chrmarti, Thanks. I tested Visual Studio Code Insider 1.30.0-insider. Found some improvement and some new problem.

The improvement is the popup (providing the device code to run in browser) appears. I can complete the Azure sign in, and see the subscription plan I have within Visual Studio Code.

However, when I try to **connect folder to experiment** , or **create workspace** , I get the following error `error: write EPROTO 101057795:error:14077419:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert access denied:d:\buildagent\workspace\318698\vendor\node\deps\openssl\openssl\ssl\s23_clnt.c:802"`

It seems like network proxy is still causing problem on some other areas. Can you take a look on this? Thanks.

**Comment by chrmarti** [comment]
@revolz Could you try opening Bash (or PowerShell) in Cloud Shell and let me know if that works? That is functionality built into the Azure Account extension.

None of the other extensions are using the new proxy support yet, that's why you see these other commands fail as before.

**Comment by revolz** [comment]
> @revolz Could you try opening Bash (or PowerShell) in Cloud Shell and let me know if that works? That is functionality built into the Azure Account extension.

@chrmarti, the Cloud Shell seems to be quite buggy at the moment.
It worked only once, that I could see the whole Azure Cloud Shell web page / command line interface. But I couldn't get it work again thereafter.
It keeps giving me the error now (after the second try and onward)
`Requesting a Cloud Shell...
Connecting terminal...
Socket error: {"errno":"EPROTO","code":"EPROTO","syscall":"write"}
Socket closed`

 
> None of the other extensions are using the new proxy support yet, that's why you see these other commands fail as before.

When can we expect the other extensions to use the new proxy support? Thanks!

**Comment by chrmarti** [comment]
@revolz I'm currently testing that new proxy support. There is no estimate yet for when this will be 'stable' and available for other extensions at the moment. (Extracted #111 for the error you report in the last comment.)

**Comment by revolz** [comment]
> Setting System Proxy to force fixed this for me:
> ![image](https://user-images.githubusercontent.com/10817275/49019956-1c270b80-f187-11e8-99d4-470b6b552c53.png)

I confirmed @newcreativegamer 's method to `set system proxy to force` is working successfully in the proxy server environment. Everything is working fine for me now. Thanks!

**Comment by chrmarti** [comment]
I have changed the default to 'force' in todays (Nov 27) Insiders build. Anyone seeing this issue please give this a try: https://code.visualstudio.com/insiders

Note that any value in the `http.proxy` setting or the HTTP_PROXY or HTTPS_PROXY environment variables will take precedence over the system's proxy setting.

**Comment by EricJizbaMSFT** [comment]
@chrmarti do other extensions have to do something to take advantage of the new proxy support in insiders? Or do they get it by default?

**Comment by chrmarti** [comment]
@EricJizbaMSFT This works transparently without any adoption. Let me know about any failure / success story.

**Comment by chrmarti** [comment]
I should add: This only works in the extension-host process. If an extension spawns another process, that process won’t get the proxy support.

**Comment by chrmarti** [comment]
@revolz Thanks, fixing it for the next insiders (likely Nov 30). (https://github.com/Microsoft/vscode/commit/84e975ef885a67e258f9a5257303a6385f4e09dd)

**Comment by revolz** [comment]
> @revolz Thanks, fixing it for the next insiders (likely Nov 30). ([Microsoft/vscode@84e975e](https://github.com/Microsoft/vscode/commit/84e975ef885a67e258f9a5257303a6385f4e09dd))

Great. Everything seems to work fine now with the latest Visual Studio Code Insiders!
Thanks for your help over the past week or two, @chrmarti. Appreciate that!

-----
Version: 1.30.0-insider (user setup)
Commit: 9e62a05bc7c5dbb74d165e06f7c3a5f3c7549096
Date: 2018-12-03T06:19:31.620Z
Electron: 2.0.12
Chrome: 61.0.3163.100
Node.js: 8.9.3
V8: 6.1.534.41
OS: Windows_NT x64 10.0.15063

**Comment by revolz** [comment]
> Setting System Proxy to force fixed this for me:
> ![image](https://user-images.githubusercontent.com/10817275/49019956-1c270b80-f187-11e8-99d4-470b6b552c53.png)

I would like to highlight, setting the Proxy System to "Force" is the only fully working option at the moment.

Setting the Proxy System to "On" only works partially for cases, even though I would expect it to work fully also.

**Comment by chrmarti** [comment]
`"On"` has the subtle difference that it will leave any `agent` passed with a single request in place and not override it, meaning you get only the support that request has set in place. `"Force"` overrides any existing [`agent` option](https://nodejs.org/api/http.html#http_http_request_options_callback).

I will rename the setting and options to clarify this.

**Comment by chrmarti** [comment]
Quick update: I have renamed the setting for tomorrow's Insiders build to `"http.proxySupport"` to better reflect its scope and it will be disabled by default due to an issue that needs more investigation (https://github.com/Microsoft/vscode/issues/60773#issuecomment-443686232). You can enable it through the settings UI (or `"http.proxySupport": "override"`).

**Comment by gstolz** [comment]
Unfortunately I also have the proxy issue here. We have a corporate proxy, which is set via pac-file by default. Tried to set the proxy manually in vscode with no success. A few months ago everything worked just fine in this setup.
If I try to interactively authenticate to Azure by running "Connect-AzureRmAccount", I get no browser window either, which works from Powershell ISE. When I run "Azure: Sign In" from the vscode extension "Azure Account" I only receive "you appear to be offline".

**Comment by chrmarti** [comment]
@gstolz Your issue is #121.

**Comment by davehax** [comment]
To anyone using Telerik Fiddler: check to see if you've set your proxy to 127.0.0.1:8888 ! If so, then you can open up Fiddler and this will allow you to sign in to Azure, or you can remove your proxy settings

**Comment by aymond** [comment]
The option for  `"http.proxySupport": "override"` worked for me.

Enterprise setup includes corporate webproxy with SSL interception. My setup includes Remote-SSH.

``` json
{
    "azureFunctions.deploySubpath": "azure\\functions",
    "azureFunctions.projectLanguage": "PowerShell",
    "azureFunctions.projectRuntime": "~3",
    "http.proxy": "http://1.2.3.4:3128",
    "http.proxySupport": "override",
    "http.proxyStrictSSL": false,
}
```

**Comment by wwlorey** [comment]
Closing in favor of https://github.com/microsoft/vscode-azure-account/issues/179
